src/focus-trap-react.js

```diff
@@ -57,8 +57,24 @@ class FocusTrap extends React.Component {
 
   /** Returns focus to the element that had focus when the trap was activated. */
   returnFocus() {
-    if (this.previouslyFocusedElement && this.previouslyFocusedElement.focus) {
-      this.previouslyFocusedElement.focus();
+    const {
+      checkCanReturnFocus,
+      onPostDeactivate,
+    } = this.props.focusTrapOptions;
+
+    const sendFocus = () => {
+      if (this.previouslyFocusedElement?.focus) {
+        this.previouslyFocusedElement.focus();
+        if (onPostDeactivate) {
+          onPostDeactivate();
```

Comment:
I'm not sure if this should be called regardless of weather `this.previouslyFocusedElement` exists or if it should only be called if focus is able to be returned.

I leaned more toward the "only call if focus is returned" option.

Reply:
Actually, I think we got this slightly wrong over in focus-trap:

https://github.com/focus-trap/focus-trap/blob/master/index.js#L548-L557
```javascript
      const finishDeactivation = () => {
        if (returnFocus) {
          delay(() => {
            tryFocus(getReturnFocusNode(state.nodeFocusedBeforeActivation));
            if (onPostDeactivate) {
              onPostDeactivate();
            }
          });
        }
      };
```

I think there's a bug there in that `onPostDeactivate()` is only called if focus is returned. Doesn't matter if focus is returned or not, the trap still got deactivated. The code should probably be this:

```javascript
      const finishDeactivation = () => {
        delay(() => {
          if (returnFocus) {
            tryFocus(getReturnFocusNode(state.nodeFocusedBeforeActivation));
          }
          if (onPostDeactivate) {
            onPostDeactivate();
          });
        }
      };
```

Meaning that onPostDeactivate() would consistently be called whenever the trap is deactivated, whether focus is returned or not, since the trap is always deactivated regardless of focus being returned...

And then it would inform this decision here: To always call it even if focus isn't returned.

👉 I can create a small PR for this over on focus-trap if you agree with my thinking.

One way or another, however, it means that because the point at which focus is returned is altered by focus-trap-react, we need to also alter the point at which onPostDeactivate() is called so it matches the focus-trap behavior.

That means that when we're building the `this.tailoredFocusTrapOptions` (see https://github.com/focus-trap/focus-trap-react/blob/master/src/focus-trap-react.js#L34), we need to set grab `props.focusTrapOptions.onPostDeactivate` and keep it in a special class member, like `this.originalPostDeactivate` and always set `onPostDeactivate: null` in `this.tailoredFocusTrapOptions` (which is what we ultimately use to create the trap at https://github.com/focus-trap/focus-trap-react/blob/master/src/focus-trap-react.js#L77)

And then it means that here, we should not do this:
```javascript
    const { checkCanReturnFocus, onPostDeactivate } = this.props.focusTrapOptions;
```
but rather this:
```javascript
    const { checkCanReturnFocus } = this.tailoredFocusTrapOptions;
    const onPostDeactivate = this.originalPostDeactivate;
```

Makes sense?

Finally, we also need to call onPostDeactivate() here if `this.returnFocusOnDeactivate` is false:
- https://github.com/focus-trap/focus-trap-react/blob/master/src/focus-trap-react.js#L109
- https://github.com/focus-trap/focus-trap-react/blob/master/src/focus-trap-react.js#L134

We should probably change the internal `returnFocus()` method to `deactivateTrap()` which does this:
```javascript
if (this.focusTrap) {
  // NOTE: we never let the trap return the focus since we do that ourselves
  this.focusTrap.deactivate({ returnFocus: false });
}

if (this.returnFocusOnDeactivate && this.previouslyFocusedElement && this.previouslyFocusedElement.focus) {
    this.previouslyFocusedElement.focus();
  }
}

if (this.originalPostDeactivate) {
  this.originalPostDeactivate.call(null); // don't call it in context of "this"
}
```

Then we can call `deactivateTrap()` from all affected locations where the trap needs to be deactivated:
- https://github.com/focus-trap/focus-trap-react/blob/master/src/focus-trap-react.js#L105
- https://github.com/focus-trap/focus-trap-react/blob/master/src/focus-trap-react.js#L129

Reply:
I'm a bit confused about the `this.tailoredFocusTrapOptions` stuff but it makes sense that you want to call `postDeactivate` regardless of weather focus is sent since the function name doesn't really say anything about returning focus. Maybe we got the name wrong. Maybe we should have called it `onReturnFocus`.

I had a look and it looks like `onPostActivate` is called regardless of weather the focus is sent or not so `onPostDeactivate` should be consistent.

Maybe we provide both `onPostDeactivate` and `onReturnFocus` options? Though I'm not sure if there is enough value in doing that to justify having the extra callback option. Probably leave it out for now.

Reply:
> I'm a bit confused about the `this.tailoredFocusTrapOptions` stuff [...]

It's because of the fact that focus-trap-react does some special handling around focus-trap's `returnFocusOnDeactivate ` because of focus issues that come-up when React is in charge of rendering. focus-trap-react sets its own flag and kind of replicates the behavior in the place of focus-trap. And because onPostDeactivate() in focus-trap comes _after_ focus is returned, we now need to specially handle that option too, so that focus-trap doesn't call onPostDeactivate() before focus-trap-react has a chance to return the focus, since the trap will actually get deactivated _before_ focus-trap-react finishes its processing. Hopefully that's a bit clearer.

> I had a look and it looks like `onPostActivate` is called regardless of weather the focus is sent or not so `onPostDeactivate` should be consistent.

OK, I will make a small PR to focus-trap to fix this.

> Maybe we provide both `onPostDeactivate` and `onReturnFocus` options? Though I'm not sure if there is enough value in doing that to justify having the extra callback option. Probably leave it out for now.

I say leave that out for now. We can add that later if there's a real need for it.